tests/cancun/eip6780_selfdestruct/test_selfdestruct.py

```diff
@@ -0,0 +1,784 @@
+        )
+        # And we store the created address for verification purposes
+        + Op.SSTORE(
+            0,
+            create_bytecode,
+        )
+    )
+    # Store the created address
+    entry_code_storage[next(storage_index)] = selfdestruct_contract_address
+
+    # Store the extcode* properties of the created address
+    current_index = next(storage_index)
+    entry_code += Op.SSTORE(
+        current_index,
+        Op.EXTCODESIZE(Op.PUSH20(selfdestruct_contract_address)),
+    )
+    entry_code_storage[current_index] = len(selfdestruct_code)
+
+    current_index = next(storage_index)
+    entry_code += Op.SSTORE(
+        current_index,
+        Op.EXTCODEHASH(Op.PUSH20(selfdestruct_contract_address)),
+    )
+    entry_code_storage[current_index] = bytes(keccak256(selfdestruct_code))
+
+    # Call the self-destructing contract multiple times as required, incrementing the wei sent each
+    # time
+    for i in range(call_times):
+        current_index = next(storage_index)
+        entry_code += Op.SSTORE(
+            current_index,
+            Op.CALL(
+                Op.GASLIMIT,  # Gas
+                Op.PUSH20(selfdestruct_contract_address),  # Address
+                i,  # Value
+                0,
+                0,
+                0,
+                0,
+            ),
+        )
+        entry_code_storage[current_index] = 1
+
+        # TODO: Why is this correct ??
+        sendall_amount += i
+
+        current_index = next(storage_index)
+        entry_code += Op.SSTORE(
+            current_index,
+            Op.BALANCE(Op.PUSH20(selfdestruct_contract_address)),
+        )
+        entry_code_storage[current_index] = 0
+
+    # Check the extcode* properties of the selfdestructing contract again
+    current_index = next(storage_index)
+    entry_code += Op.SSTORE(
+        current_index,
+        Op.EXTCODESIZE(Op.PUSH20(selfdestruct_contract_address)),
+    )
+    entry_code_storage[current_index] = len(selfdestruct_code)
+
+    current_index = next(storage_index)
+    entry_code += Op.SSTORE(
+        current_index,
+        Op.EXTCODEHASH(Op.PUSH20(selfdestruct_contract_address)),
+    )
+    entry_code_storage[current_index] = bytes(
+        keccak256(selfdestruct_code if call_times > 0 else b"")
+    )
+
+    # Lastly return "0x00" so the entry point contract is created and we can retain the stored
+    # values for verification.
+    entry_code += Op.RETURN(len(selfdestruct_contract_initcode), 1)
+
+    if selfdestruct_contract_initial_balance > 0:
+        # Address where the contract is created already had some balance,
+        # which must be included in the send-all operation
+        sendall_amount += selfdestruct_contract_initial_balance
+
+    post: Dict[str, Account] = {
+        entry_code_address: Account(
+            code="0x00",
+            storage=entry_code_storage,
+        ),
+        selfdestruct_contract_address: Account.NONEXISTENT,  # type: ignore
```

Comment:
For some cases of cancun this wiill fail.  Here's what I have (I have Cancun wired in instead of plus notation).  If the account is pre-funded it will sweep out balance, but storage and nonce remains.

```suggestion
        selfdestruct_contract_address: Account(balance=0, storage={0:1}, nonce=1) if is_fork(fork=fork, which=Cancun) & prefunded_selfdestruct_address else  Account.NONEXISTENT,
```

Reply:
Other than this, my besu impl passes all test (https://github.com/hyperledger/besu/pull/5430)

Reply:
This case is fairly tricky.  Account exists prior to transaction start.  Create adds the new code to the existing code.  Because the contract creates a new contract a nonce increment occurs.  

Is this a contract create when done into an account with a balance?

- So we need semantic clarification if the phrase `contract calling SELFDESTRUCT was created` - is it created when the CREATE operation returns or is it created when ether is deposited in the account? Or are account creation and contract creation two distinct events within the spec? Either is fine, but both can be argued from the spec with no clear winner.

Reply:
Yes, I think that was the point of the tests where there's a previous balance in the address.

I think it would be a huge problem if someone could inhibit the ability to self-destruct by simply sending 1 wei before the transaction creating the contract.

What's besu's current implementation? These tests should pass if the account can self-destruct even when there was balance beforehand.

Reply:
Besu's initial implementation was to track account creation.  I just checked in code to only consider the results of CREATE/CREATE2 and the address of a transaction in create mode.  So if you self-destruct in your root-level init code there is no footprint left.

Reply:
PR to clarify the two cases I felt were unclear

https://github.com/ethereum/EIPs/pull/7308